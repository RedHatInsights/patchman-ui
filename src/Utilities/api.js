import { encodeApiParams } from './Helpers';

export function createApiCall(
    endpoint,
    method,
    parameters = undefined,
    data = undefined
) {
    if (parameters && method === 'get') {
        endpoint = endpoint.concat(encodeApiParams(parameters));
    }

    let result = window.insights.chrome.auth
    .getUser()
    .then(() =>
        fetch('/api/patch/v1' + endpoint, {
            method,
            credentials: 'include',
            body: JSON.stringify(data)
        })
    )
    .then(res => {
        if (!res.ok) {
            const contentType = res.headers.get('content-type');
            if (contentType.indexOf('json') !== -1) {
                throw res;
            } else {
                throw {
                    errors: [
                        { status: res.status, detail: res.statusText }
                    ]
                };
            }
        }

        return res.json();
    })
    .catch(caughtError => {
        const error = Promise.resolve(caughtError || {});
        const genericError = {
            title:
                    'There was an error getting data'
        };
        return error.then(error => {
            const [info] = error.errors;
            const result = error.errors && { ...genericError, detail: info.detail, status: info.status } || genericError;
            throw result ;
        });
    });
    return result;
}

export const fetchApplicableAdvisoriesApi = params => {
    return createApiCall('/advisories', 'get', params);
};

export const fetchApplicableSystemAdvisoriesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/advisories`, 'get', allParams);
};

export const fetchSystems = params => {
    return createApiCall('/systems', 'get', params);
};

export const fetchAdvisoryDetailsApi = params => {
    return createApiCall(`/advisories/${params.advisoryName}`, 'get');
};

export const fetchApplicablePackagesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/packages`, 'get', allParams);
};

export const fetchAffectedSystems = params => {
    const { id, ...args } = params;
    return createApiCall(`/advisories/${id}/systems`, 'get', args);
};

export const fetchPackagesList = params => {
    return createApiCall('/packages', 'get', params);
};

export const exportAdvisoriesCSV = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportAdvisoriesJSON = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};

export const exportSystemsCSV = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportSystemsJSON = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};
