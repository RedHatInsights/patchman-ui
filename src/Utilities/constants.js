import BugIcon from '@patternfly/react-icons/dist/js/icons/bug-icon';
import EnhancementIcon from '@patternfly/react-icons/dist/js/icons/enhancement-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import React from 'react';
import { subtractDate } from './Helpers';

export const STATUS_REJECTED = 'rejected';
export const STATUS_LOADING = 'loading';
export const STATUS_RESOLVED = 'resolved';

export const compoundSortValues = {
    operating_system: {
        asc: 'os_name,os_major,os_minor',
        desc: '-os_name,-os_major,-os_minor'
    },
    applicable_advisories: {
        asc: 'rhsa_count,rhba_count,rhea_count',
        desc: '-rhsa_count,-rhba_count,-rhea_count'
    }
};

// messy check because of frontend-components tests
export const ENABLE_PACKAGES = window?.insights?.chrome?.isBeta && insights?.chrome?.isBeta();
export const storeListDefaults = {
    rows: [],
    metadata: {
        limit: 25,
        offset: 0,
        total_items: 0
    },
    expandedRows: {},
    selectedRows: {},
    queryParams: {},
    error: {}
};

export const publicDateOptions = [
    { apiValue: '', label: 'All', value: 'all' },
    {
        apiValue: `gt:${subtractDate(7)}`,
        label: 'Last 7 days',
        value: 'last7'
    },
    {
        apiValue: `gt:${subtractDate(30)}`,
        label: 'Last 30 days',
        value: 'last30'
    },
    {
        apiValue: `gt:${subtractDate(90)}`,
        label: 'Last 90 days',
        value: 'last90'
    },
    {
        apiValue: `gt:${subtractDate(365)}`,
        label: 'Last year',
        value: 'lastYear'
    },
    {
        apiValue: `lt:${subtractDate(365)}`,
        label: 'More than 1 year ago',
        value: 'moreThanYear'
    }
];

export const advisorySeverities = [
    {
        value: 0,
        label: 'N/A',
        color: 'var(--pf-global--Color--200)'
    },
    {
        value: 1,
        label: 'Low',
        color: 'var(--pf-global--Color--200)',
        text:
            'This rating is given to all other issues that have a security impact. These are the types of vulnerabilities' +
            ' that are believed to require unlikely circumstances to be able to be exploited, or where a successful exploit' +
            ' would give minimal consequences.'
    },
    {
        value: 2,
        label: 'Moderate',
        color: 'var(--pf-global--warning-color--100)',
        text:
            'This rating is given to flaws that may be more difficult to exploit but could still lead to some' +
            ' compromise of the confidentiality, integrity, or availability of resources, under certain circumstances.' +
            ' These are the types of vulnerabilities that could have had a Critical impact or Important impact' +
            ' but are less easily exploited based on a technical evaluation of the flaw, or affect unlikely' +
            ' configurations.'
    },
    {
        value: 3,
        label: 'Important',
        color: '#ec7a08',
        text:
            'This rating is given to flaws that can easily compromise the confidentiality, integrity, or availability' +
            ' of resources. These are the types of vulnerabilities that allow local users to gain privileges, allow' +
            ' unauthenticated remote users to view resources that should otherwise be protected by authentication,' +
            ' allow authenticated remote users to execute arbitrary code, or allow remote users to cause a denial' +
            ' of service.'
    },
    {
        value: 4,
        label: 'Critical',
        color: 'var(--pf-global--danger-color--100)',
        text:
            'This rating is given to flaws that could be easily exploited by a remote unauthenticated attacker' +
            ' and lead to system compromise (arbitrary code execution) without requiring user interaction.' +
            ' These are the types of vulnerabilities that can be exploited by worms. Flaws that require an authenticated' +
            ' remote user, a local user, or an unlikely configuration are not classed as Critical impact.'
    }
];

export const advisoryTypes = [
    {
        value: 3,
        label: 'Security',
        icon: <SecurityIcon />
    },
    {
        value: 2,
        label: 'Bugfix',
        icon: <BugIcon />
    },
    {
        value: 1,
        label: 'Enhancement',
        icon: <EnhancementIcon />
    }
];

export const updatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: true,
        label: 'Upgradable'
    }
];

export const packagesListUpdatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: 'gt:0',
        label: 'Upgradable'
    }
];

export const filterCategories = {
    advisory_type: {
        label: 'Advisory type',
        values: advisoryTypes
    },
    public_date: {
        label: 'Public date',
        values: publicDateOptions
    },
    updatable: {
        label: 'Status',
        values: updatableTypes
    },
    systems_updatable: {
        label: 'Status',
        values: packagesListUpdatableTypes
    }
};

export const entityTypes = {
    advisories: 'advisories',
    packages: 'packages'
};

export const ReadOnlyNotification = {
    title: 'title',
    detail: 'message'
};

export const remediationIdentifiers = {
    package: 'patch-package',
    advisory: 'patch-advisory'
};
