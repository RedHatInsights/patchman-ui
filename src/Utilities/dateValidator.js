import { intl } from './IntlProvider';
import messages from '../Messages';

/*
validates date of a type YYYY-MM-DD with starting limit 1990-01-01.
regex is used to comply the Patternfly date format.
*/
const dateValidator = (dateStr) => {
    const regex = /^(\d{4})-(0[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])$/;

    if (dateStr && typeof dateStr === 'string' && dateStr.match(regex) === null) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    const date = new Date(dateStr);
    const timestamp = date.getTime();

    //Month had to be set to equal 0 to get the first month of the year.
    const minDate = new Date(1990, 0, 1);

    if (typeof timestamp !== 'number' || Number.isNaN(timestamp)) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    if (date < minDate) {
        return intl.formatMessage(messages.labelsErrorDateLimit);
    }

    return;
};

export default dateValidator;
