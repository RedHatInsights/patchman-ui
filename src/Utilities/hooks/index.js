export * from './Hooks';
export * from './useFeatureFlag';
export * from './useOnSelect';
export * from './usePatchSetState';
export { useRemediationDataProvider } from './useRemediationDataProvider';
export * from './useFetchBatched';
