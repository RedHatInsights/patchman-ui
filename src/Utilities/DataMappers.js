import { processDate } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/helpers';
import { flatMap } from 'lodash';
import React from 'react';
import AdvisoryType from '../PresentationalComponents/AdvisoryType/AdvisoryType';
import { EmptyAdvisoryList, EmptyPackagesList } from '../PresentationalComponents/Snippets/EmptyStates';
import { entityTypes } from './constants';
import { createUpgradableColumn, handlePatchLink } from './Helpers';
import { DescriptionWithLink } from '../PresentationalComponents/Snippets/DescriptionWithLink';

export const createAdvisoriesRows = (rows, expandedRows, selectedRows) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        {
                            title: handlePatchLink(
                                entityTypes.advisories,
                                row.id,
                                row.attributes.applicable_systems
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 5 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemAdvisoriesRows = (
    rows,
    expandedRows,
    selectedRows
) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 6 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                packages_installed: row.attributes.packages_installed,
                applicable_advisories: [
                    row.attributes.rhea_count || 0,
                    row.attributes.rhba_count || 0,
                    row.attributes.rhsa_count || 0
                ],
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createPackageSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                installed_evra: row.installed_evra,
                available_evra: row.updatable && row.available_evra || row.installed_evra,
                upgradable: row.updatable,
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createSystemPackagesRows = (rows, selectedRows = {}) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            const pkgUpdates = pkg.updates || [];
            const latestUpdate = pkgUpdates[pkgUpdates.length - 1];

            return {
                id: pkg.name,
                key: pkg.name,
                selected: selectedRows[pkg.name] !== undefined,
                disableCheckbox: !latestUpdate,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.evra },
                    { title: (latestUpdate && latestUpdate.evra) || pkg.evra },
                    { title: createUpgradableColumn(pkg.updatable) },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};

export const createPackagesRows = (rows) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            return {
                id: pkg.name,
                key: pkg.name,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.systems_installed },
                    { title: pkg.systems_updatable },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};
