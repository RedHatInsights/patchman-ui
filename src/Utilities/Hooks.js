import { SortByDirection } from '@patternfly/react-table/dist/js';
import React from 'react';
import { APPLICABLE_ADVISORIES_ASC, APPLICABLE_ADVISORIES_DESC } from './constants';
import { convertLimitOffset, getLimitFromPageSize, getOffsetFromPageLimit } from './Helpers';
import isDeepEqualReact from 'fast-deep-equal/react';

export const useSetPage = (limit, callback) => {
    const onSetPage = React.useCallback((_, page) =>
        callback({ offset: getOffsetFromPageLimit(page, limit) })
    );
    return onSetPage;
};

export const useHandleRefresh = (metadata, callback) => {
    const handleRefresh = React.useCallback(({ page, per_page: perPage }) => {
        const offset = getOffsetFromPageLimit(page, perPage);
        const limit = getLimitFromPageSize(perPage);
        (metadata.offset !== offset || metadata.limit !== limit) &&
            callback({
                ...(metadata.offset !== offset && { offset }),
                ...(metadata.limit !== limit && { limit })
            });
    });
    return handleRefresh;
};

export const usePagePerPage = (limit, offset) => {
    const [page, perPage] = React.useMemo(
        () => convertLimitOffset(limit, offset),
        [limit, offset]
    );
    return [page, perPage];
};

export const usePerPageSelect = callback => {
    const onPerPageSelect = React.useCallback((_, perPage) =>
        callback({ limit: getLimitFromPageSize(perPage), offset: 0 })
    );
    return onPerPageSelect;
};

export const useSortColumn = (columns, callback, offset = 0) => {
    const onSort = React.useCallback((_, index, direction) => {
        let columnName = columns[index - offset].key;
        if (direction === SortByDirection.desc) {
            columnName = '-' + columnName;
        }

        if (columnName === '-applicable_advisories') {
            columnName = APPLICABLE_ADVISORIES_DESC;
        }
        else if (columnName === 'applicable_advisories') {
            columnName = APPLICABLE_ADVISORIES_ASC;
        }

        callback({ sort: columnName });
    });
    return onSort;
};

export const useRemoveFilter = (filters, callback) => {
    const removeFilter = React.useCallback((event, selected) => {
        let newParams = { filter: {} };
        selected.forEach(selectedItem => {
            let { id: categoryId, chips } = selectedItem;
            if (categoryId !== 'search') {
                let activeFilter = filters[categoryId];
                const toRemove = chips.map(item => item.id.toString());
                if (Array.isArray(activeFilter)) {
                    newParams.filter[categoryId] = activeFilter.filter(
                        item => !toRemove.includes(item.toString())
                    );
                } else {
                    newParams.filter[categoryId] = '';
                }
            } else {
                newParams.search = '';
            }
        });
        callback({ ...newParams });
    });
    return removeFilter;
};

export const useOnSelect = (rawData, selectedRows, fetchAllData, selectRows, constructFilename = undefined) =>{

    const onSelect =  React.useCallback((event, selected, rowId) => {
        const createSelectedRow = (rawData, toSelect = []) => {
            rawData.forEach((row)=>{
                toSelect.push(
                    {
                        id: row.id && row.id || row.name,
                        selected: constructFilename && constructFilename(row) || row.id
                    }
                );});

            return toSelect;
        };

        switch (event) {
            case 'none': {
                const toSelect = [];
                Object.keys(selectedRows).forEach(id=>{
                    toSelect.push(
                        {
                            id,
                            selected: false
                        }
                    );
                });
                selectRows(toSelect);
                break;
            }

            case 'page': {
                selectRows(createSelectedRow(rawData));
                break;
            }

            case 'all': {
                const fetchCallback = ({ data }) => {
                    selectRows(createSelectedRow(data));
                };

                fetchAllData().then(fetchCallback);

                break;
            }

            default: {
                selectRows([{
                    id: rawData[rowId].id && rawData[rowId].id || rawData[rowId].name,
                    selected: selected && (constructFilename && constructFilename(rawData[rowId]) || true)
                }]);
            }

        }}
    );

    return onSelect;
};

export const setPageTitle = (title) => {
    React.useEffect(() => {
        if (title) {
            document.title = `${title} - Patch | Red Hat Insights`;
        }
    }, [title]);
};

export const useDeepCompareEffect = (effect, deps) => {
    const ref = React.useRef(undefined);

    if (!ref.current || !isDeepEqualReact(deps, ref.current)) {
        ref.current = deps;
    }

    React.useEffect(effect, ref.current);
};
