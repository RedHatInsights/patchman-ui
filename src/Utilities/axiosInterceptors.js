import {
    authInterceptor,
    responseDataInterceptor
} from '@redhat-cloud-services/frontend-components-utilities/files/interceptors';
import axios from 'axios';
import { ReadOnlyNotification } from './constants';
const axiosInstance = axios.create();

export async function  accessCheckInterceptor (config)  {

    const userPermissions = await insights.chrome.getUserPermissions();
    const inventoryPermissionList = ['inventory:*:*', 'inventory:*:read', 'inventory:hosts:read'];

    const hasInventoryAccess = userPermissions.some((access) => inventoryPermissionList.includes(access?.permission));

    if (!hasInventoryAccess) {
        throw new axios.Cancel('Operation canceled. User does not have access to Inventory');
    }

    return config;
}

export function errorInterceptor(err) {

    if (!axios.isCancel(err)) {

        const { response, isAxiosError } = { ...err };

        if (response && isAxiosError) {
            const { status, statusText, data } = response;

            if (!status) {
                throw err;
            } else {
                const genericError = {
                    title:
                            'There was an error getting data'
                };

                const result = { ...genericError, detail: data.error && data.error || statusText, status };

                throw result;
            }
        }

        throw err;
    }
}

export function readOnlyInterceptor(error) {
    if (error.response && error.response.status === 503) {
        const data = ReadOnlyNotification;
        throw data;
    }

    throw error;
}

axiosInstance.interceptors.request.use(accessCheckInterceptor);
axiosInstance.interceptors.request.use(authInterceptor);
axiosInstance.interceptors.response.use(responseDataInterceptor);
axiosInstance.interceptors.response.use(null, readOnlyInterceptor);
axiosInstance.interceptors.response.use(null, errorInterceptor);

export default axiosInstance;
