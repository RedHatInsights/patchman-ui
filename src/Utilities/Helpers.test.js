/* eslint-disable */
import { SortByDirection } from '@patternfly/react-table/dist/js';
import toJson from 'enzyme-to-json';
import { publicDateOptions } from '../Utilities/constants';
import { addOrRemoveItemFromSet, arrayFromObj, buildFilterChips, changeListParams, convertLimitOffset, createAdvisoriesIcons, createSortBy, decodeQueryparams, encodeApiParams, encodeParams, encodeURLParams, getFilterValue, getLimitFromPageSize, getNewSelectedItems, getOffsetFromPageLimit, getRowIdByIndexExpandable, getSeverityById, handlePatchLink, remediationProvider } from './Helpers';

const TestHook = ({ callback }) => {
    callback();
    return null;
};

export const testHook = callback => {
    mount(<TestHook callback={callback} />);
};

describe('Helpers tests', () => {
    let header = [
        {
            key: 'a'
        },
        {
            key: 'b'
        },
        {
            key: 'c'
        },
        {
            key: 'd'
        }
    ];
    it.each`
    header      | value         | offset    | result
    ${header}   | ${['-a']}     | ${0}      | ${{ index: 0, direction: SortByDirection.desc }}
    ${header}   | ${['c']}      | ${1}      | ${{ index: 3, direction: SortByDirection.asc }}
    ${undefined}   | ${undefined}      | ${undefined}      | ${{}}
    `('createSortBy: Should create correct sort for $value and offset $offset', ({header, value, offset, result}) => {
        let ret = createSortBy(header, value, offset);
        expect(ret).toEqual(result);
    });

    it.each`
    rhea | rhba | rhsa
    ${1} | ${2} | ${3}
    ${0} | ${0} | ${0}
    ${0} | ${5} | ${3}
    ${1} | ${0} | ${3}
    ${1} | ${2} | ${0}
    `('createAdvisoriesIcons: Should match advisory icons snapshot for [$rhea, $rhba, $rhsa]', ({rhea, rhba, rhsa}) => {
        let wrapper = shallow(createAdvisoriesIcons([rhea, rhba, rhsa]));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it.each`
    severity | result
    ${123}   | ${"N/A"}
    ${0}     | ${"N/A"}
    ${1}     | ${"Low"}
    ${2}     | ${"Moderate"}
    ${3}     | ${"Important"}
    ${4}     | ${"Critical"}
    `('getSeverityById: Should match $severity to value $result', ({severity, result}) => {
        expect(getSeverityById(severity).label).toEqual(result);
    });

    it('addOrRemoveItemFromSet: Should create correct object  ', () => {
        let inputArr = [
            {
                rowId: 0,
                value: 'a'
            },
            {
                rowId: 1,
                value: 'b'
            },
            {
                rowId: 2,
                value: 'c'
            },
            {
                rowId: 3,
                value: 'd'
            },
            {
                rowId: 5,
                value: undefined
            }
        ];
        let targetObj = { '4': 'e' };
        let expected = { '0': 'a', '1': 'b', '2': 'c', '3': 'd', '4': 'e' };
        let ret = addOrRemoveItemFromSet(targetObj, inputArr);

        expect(ret).toEqual(expected);
    });

    it('convertLimitOffset: Should get correct limit and offset', () => {
        let ret = convertLimitOffset(50, 1100);
        let expected = [23, 50];

        expect(ret).toEqual(expected);
    });

    it('getRowIdByIndexExpandable: Should return correct id', () => {
        let inputArr = [
            {
                id: 0,
                value: 'a'
            },
            {
                id: 1,
                value: 'b'
            },
            {
                id: 2,
                value: 'c'
            },
            {
                id: 3,
                value: 'd'
            }
        ];
        let ret = getRowIdByIndexExpandable(inputArr, 2);
        let expected = 1;

        expect(ret).toEqual(expected);
    });

    it('handlePatchLink: Get advisory link from inventory', () => {
        delete global.window.location;
        global.window.location = {
            href: 'https://cloud.redhat.com/insights/inventory'
        };
        const host = document.baseURI;
        let advisoryName = 'ABCD';
        const expected = `${host}insights/patch/advisories/${advisoryName}`;
        let result = handlePatchLink('advisories',advisoryName);
        let {
            props: { href, children }
        } = result;
        expect(href).toEqual(expected);
        expect(children).toEqual(advisoryName);
    });

    it('handlePatchLink: Get advisory link from inventory with custom text', () => {
        delete global.window.location;
        global.window.location = {
            href: 'https://cloud.redhat.com/insights/inventory'
        };
        const host = document.baseURI;
        let advisoryName = 'ABCD';
        const expected = `${host}insights/patch/advisories/${advisoryName}`;
        let result = handlePatchLink('advisories',advisoryName, 'custom text');
        let {
            props: { href, children }
        } = result;
        expect(href).toEqual(expected);
        expect(children).toEqual('custom text');
    });

    it('handlePatchLink: Get advisory link from patch', () => {
        delete global.window.location;
        global.window.location = {
            href: 'https://cloud.redhat.com/rhel/patch'
        };
        let advisoryName = 'ABCD';
        const expected = `/advisories/${advisoryName}`;
        let result = handlePatchLink('advisories',advisoryName);
        let {
            props: { to, children }
        } = result;
        expect(to).toEqual(expected);
        expect(children).toEqual(advisoryName);
    });

    it('handlePatchLink: Get advisory link from patch', () => {
        delete global.window.location;
        global.window.location = {
            href: 'https://cloud.redhat.com/rhel/patch'
        };
        let advisoryName = 'ABCD';
        const expected = `/advisories/${advisoryName}`;
        let result = handlePatchLink('advisories',advisoryName, 'custom text');
        let {
            props: { to, children }
        } = result;
        expect(to).toEqual(expected);
        expect(children).toEqual('custom text');
    });

    it('getLimitFromPageSize: Should get correct Limit', () => {
        let limit = Math.random();
        let result = getLimitFromPageSize(limit);
        expect(result).toEqual(limit);
    });

    it('getOffsetFromPageLimit: Should get correct offset', () => {
        let page = Math.random() * 10;
        let limit = Math.random() * 10;
        let result = getOffsetFromPageLimit(page, limit);
        expect(result).toEqual(page * limit - limit);
    });

    it('arrayFromObj: Should create correct array', () => {
        let items = {a: "value1", b: false, c: "value2"}
        let expected = ["value1", "value2"]
        let result = arrayFromObj(items);
        expect(result).toEqual(expected);
    });

    it.each`
    issues         | systems            | result
    ${["issue-1"]} | ${["system-1"]}    | ${{issues: [{id: "patch-advisory:issue-1", description: "issue-1"}],systems: ["system-1"]}}
    ${"issue-1"}   | ${"system-1"}      | ${{issues: [{id: "patch-advisory:issue-1", description: "issue-1"}],systems: ["system-1"]}}
    ${[]}          | ${["system-1"]}    | ${false}
    `('remediationProvider: Should create correct remediation object for $issues $systems', ({issues, systems, result}) => {
        expect(remediationProvider(issues,systems)).toEqual(result);
    });

    it.each`
    category         | key                  | result
    ${"public_date"} | ${"last7"}           | ${publicDateOptions[1]}
    ${"public_date"} | ${"random value"}    | ${{apiValue: "random value"}}
    ${undefined}     | ${"last7"}           | ${undefined}
    `('getFilterValue: Should create object for $category', ({category, key, result}) => {
        expect(getFilterValue(category,key)).toEqual(result);
    });

    it.each`
    parameters                                         | shouldTranslate   | result
    ${{search: "trolo"}}                               | ${true}           | ${"?search=trolo"}
    ${{search: ""}}                                    | ${true}           | ${"?"}
    ${{filter: {advisory_type: 2}}}                    | ${false}          | ${"?filter%5Badvisory_type%5D=2"}
    ${{filter: {advisory_type: [1,2]}}}                | ${true}           | ${"?filter%5Badvisory_type%5D=in%3A1%2C2"}
    ${{filter: {advisory_type: [1,2]}, param: "text"}} | ${true}           | ${"?param=text&filter%5Badvisory_type%5D=in%3A1%2C2"}
    `('encodeParams: Should encode parameters $parameters', ({parameters, shouldTranslate, result}) => {
        expect(encodeParams(parameters,shouldTranslate)).toEqual(result);
    });

    it.each`
    parameters                                              | result
    ${"search=trolo"}                                       | ${{search: "trolo"}}
    ${"filter%5Badvisory_type%5D=2"}                        | ${{filter: {advisory_type: "2"}}}
    ${"param=text&filter%5Badvisory_type%5D=in%3A1%2C2"}    | ${{filter: {advisory_type: ["1","2"]}, param: "text"}}
    `('decodeQueryparams: Should decodeQueryParams $parameters', ({parameters, result}) => {
        expect(decodeQueryparams(parameters)).toEqual(result);
    });

    it.each`
    filters               | search       | result
    ${{advisory_type: 2}} | ${undefined} | ${[{"category": "Advisory type", "chips": [{"id": 2, "name": "Bugfix", "value": 2}], "id": "advisory_type"}]}
    ${undefined}          | ${"firefox"} | ${[{"category": "Search", "chips": [{"name": "firefox", "value": "firefox"}], "id": "search"}]}
     `('buildFilterChips: Should build correct filter chip, $filters, $search ', ({filters, search, result}) => {
        expect(buildFilterChips(filters, search)).toEqual(result);
    });

    it.each`
    oldParams               | newParams       | result
    ${{param: "Hey!"}}      | ${{param: "Yo!"}} | ${{param: "Yo!"}}
    ${{offset: 15}} | ${{limit:100}} | ${{"limit": 100, "offset": 0}}
    ${{filter: {advisory_type: 2}}} | ${{filter: {public_date: "last7" }}} | ${{filter:{advisory_type: 2, public_date: "last7"}, offset: 0}}
     `('changeListParams: Should return correct parameters', ({oldParams, newParams, result}) => {
        expect(changeListParams(oldParams, newParams)).toEqual(result);
    });

    it.each`
    parameters                | result
    ${{filter: {advisory_type: 2}}}  | ${"?filter%5Badvisory_type%5D=2"}
    ${{filter: {advisory_type: 1}, id: 15}}  | ${"?filter%5Badvisory_type%5D=1"}
     `('encodeURLParams: Should return correct string for $parameters', ({parameters, result}) => {
        expect(encodeURLParams(parameters)).toEqual(result);
    });

    it.each`
    parameters                | result
    ${{filter: {advisory_type: 2}}}  | ${"?filter%5Badvisory_type%5D=2"}
    ${{filter: {advisory_type: 1}, id: 15}}  | ${"?id=15&filter%5Badvisory_type%5D=1"}
     `('encodeApiParams: Should return correct string for $parameters', ({parameters, result}) => {
        expect(encodeApiParams(parameters)).toEqual(result);
    });

    it.each`
    selectedItems                                                            | currentItems           | result
    ${ { selectedItems: {id: "a", selected: true} , selectionType: 'all' }}  | ${{c: true, d: false}} | ${{"a": true, "c": true, "d": false}}
     `('getNewSelectedItems: Should return new set of selected items', ({selectedItems, currentItems,result}) => {
        expect(getNewSelectedItems(selectedItems, currentItems)).toEqual(result);
    });
});
/* eslint-enable */
