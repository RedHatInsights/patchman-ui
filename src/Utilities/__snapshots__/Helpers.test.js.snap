// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers tests createAdvisoriesIcons: Should match advisory icons snapshot for [0, 0, 0] 1`] = `
<div>
  No applicable advisories
</div>
`;

exports[`Helpers tests createAdvisoriesIcons: Should match advisory icons snapshot for [0, 5, 3] 1`] = `
<div>
  <AdvisoriesIcon
    Icon={[Function]}
    count={3}
    tooltipText="Security advisories"
  />
  <AdvisoriesIcon
    Icon={[Function]}
    count={5}
    tooltipText="Bug fixes"
  />
</div>
`;

exports[`Helpers tests createAdvisoriesIcons: Should match advisory icons snapshot for [1, 0, 3] 1`] = `
<div>
  <AdvisoriesIcon
    Icon={[Function]}
    count={3}
    tooltipText="Security advisories"
  />
  <AdvisoriesIcon
    Icon={[Function]}
    count={1}
    tooltipText="Enhancements"
  />
</div>
`;

exports[`Helpers tests createAdvisoriesIcons: Should match advisory icons snapshot for [1, 2, 0] 1`] = `
<div>
  <AdvisoriesIcon
    Icon={[Function]}
    count={2}
    tooltipText="Bug fixes"
  />
  <AdvisoriesIcon
    Icon={[Function]}
    count={1}
    tooltipText="Enhancements"
  />
</div>
`;

exports[`Helpers tests createAdvisoriesIcons: Should match advisory icons snapshot for [1, 2, 3] 1`] = `
<div>
  <AdvisoriesIcon
    Icon={[Function]}
    count={3}
    tooltipText="Security advisories"
  />
  <AdvisoriesIcon
    Icon={[Function]}
    count={2}
    tooltipText="Bug fixes"
  />
  <AdvisoriesIcon
    Icon={[Function]}
    count={1}
    tooltipText="Enhancements"
  />
</div>
`;
